### ISSUE TITLE
Edit Context: Screen Reader Users Feedback

### ISSUE THREAD

**Issue by aiday-mar** [microsoft/vscode#229051]
Pinging @jooyoungseo, @rperez030 and @meganrogge 

Recently we have been working on adopting the EditContext API (https://developer.mozilla.org/en-US/docs/Web/API/EditContext_API) within VS Code. The EditContext is a new DOM property that can be set on DOM elements which decouples text input from the textual mutations of the DOM element. Essentially when the user focuses a DOM element on which an EditContext is set and types, the EditContext fires 'textupdate' events, and it is up to the user to mutate the DOM element with the changes from this event. 

There are several reasons why we have adopted this API:
- This API has allowed us to greatly simplify the code which handles text input events.
- This API has allowed us to close numerous IME related bugs
- This API can allow us to emit customized typing information from the input events 

We have an experimental setting which enables the EditContext API with ID `editor.experimentalEditContextEnabled`. We would like to ask @jooyoungseo and @rperez030 if when you have time you could try the setting with a screen reader and let us know if you see any issues on typing in the various inputs of VS Code (could be the editor, the panel chat input, the SCM view, the quick input, ie any input that accepts text insertions). We would like to gather feedback from screen reader users before enabling this setting by default.

**Comment by jooyoungseo** [comment]
NVDA does not seem to support this new editor mode; word-by-word navigation with JAWS is not reliable. Sorry for the brevity. This is my initial and quick testing results.

**Comment by rperez030** [comment]
Hi team,

After enabling the setting, both NVDA and Narrator no longer recognize the editor as a text editor and cannot track the cursor properly. While NVDA performs slightly better, it is still not usable.

Based on my observations, this may be due to Chromium not returning the correct roles and states through the IAccessible API.

Here are the API returns with the setting enabled and disabled:

#### With the Setting Enabled:
- **IAccessible:**
  - **accName:** 'This is a test. • Untitled-1'
  - **accRole:** ROLE_SYSTEM_GROUPING
  - **accState:** STATE_SYSTEM_FOCUSED, STATE_SYSTEM_FOCUSABLE, STATE_SYSTEM_VALID (1048580)
  - **accDescription:** None
  - **accValue:** 'This is a test.\n\nThis is a line.\n\nThis is a line.'

- **IAccessible2:**
  - **windowHandle:** 527658
  - **uniqueID:** -2569
  - **role:** IA2_ROLE_SECTION
  - **states:** IA2_STATE_EDITABLE, IA2_STATE_OPAQUE, IA2_STATE_SELECTABLE_TEXT, IA2_STATE_SINGLE_LINE (13320)

#### With the Setting Disabled:
- **IAccessible:**
  - **accName:** 'This is a test. • Untitled-1'
  - **accRole:** ROLE_SYSTEM_TEXT
  - **accState:** STATE_SYSTEM_FOCUSED, STATE_SYSTEM_FOCUSABLE, STATE_SYSTEM_VALID (1048580)
  - **accDescription:** None
  - **accValue:** 'This is a test.\n\nThis is a line.\n\nThis is a line.'

- **IAccessible2:**
  - **windowHandle:** 527658
  - **uniqueID:** -12173
  - **role:** ROLE_SYSTEM_TEXT
  - **states:** IA2_STATE_EDITABLE, IA2_STATE_MULTI_LINE, IA2_STATE_OPAQUE, IA2_STATE_SELECTABLE_TEXT, IA2_STATE_SUPPORTS_AUTOCOMPLETION (38408)

I think this will require some work in Chromium to be properly supported.

**Comment by aiday-mar** [comment]
Hi @rperez030 and @jooyoungseo thank you both for your testing. Apologies I should have tested myself first with NVDA to catch this simple bug. I will have a look for a fix

**Comment by aiday-mar** [comment]
Hi @rperez030 I am using NVDA with the setting `editor.experimentalEditContextEnabled`. I put the cursor in the editor and am using the arrow keys to read the lines one by one, this seems to work. Reading letter by letter seems to work too. 

Could I ask you what you mean exactly by NVDA no longer recognizes the editor as a text editor and that it can not track the cursor properly, and potentially provide steps to reproduce this issue?

**Comment by rperez030** [comment]
Open VS code, and write the following in a new document.

This is a test.

This is a line.

This is another line.

Use the arrow keys to navigate, line by line and character by character. Observe how NVDA reports the empty lines between the text lines.

Now, enable editor.experimentalEditContextEnabled, and repeat the test.

At least in my environment, NVDA no longer tracks the cursor, and the entire document appears as a single line in the Braille display.

**Comment by aiday-mar** [comment]
Hi @rperez030 thank you for the clarification. Today I am testing the Edit Context with NVDA with the recent aria attributes changes.

After setting the setting to true, reloading the window, and after some (I assume notification) text is read, I place the cursor at the top of the editor and hear the following as I move the cursor down:

`This is a test`,  `Blank`, `This is a line`, `Blank`, `This is another line`

This seems to correspond to what is read by NVDA without the setting.

The text that is read however does not appear in the NVDA speech viewer, nor do I see it in the Braille Viewer. I tested with the setting turned off, and the text does not appear there either. I am not sure if I am missing some config to see the text that would be sent to a braille device. 

Please let me know if I am missing or misunderstanding something.

**Comment by rperez030** [comment]
In theory, everything that gets sent to the speech subsystem should also be available in the speech viewer. The same applies to braille. does the speech viewer even work for you outside VS Code? Happy to get on a call to troubleshoot if that helps speed things up.

my schedule is quite busy today, but we could connect tomorrow or friday.

**Comment by aiday-mar** [comment]
Hi @rperez030 thank you for suggesting. Today I will not be able to, but tomorrow or after tomorrow I can. What time would suit you? Thanks again.

**Comment by aiday-mar** [comment]
Hi @rperez030. After updating VS Code I am able to see the correct text read out in the speech viewer on NVDA and the braille display. The blank lines between the text are read out by NVDA as `blank`. Please let me know if you notice other accessibility issues in the editor with the experimental setting.

**Comment by jooyoungseo** [comment]
@aiday-mar -- Have you tested word-by-word navigation thoroughly? When I tested the following with ctrl+LeftArrow and ctrl+RightArrow multiple times, the cursor tracking is not mapped with what is being heard.

> The quick brown fox jumps over the lazy dog.

1. Press Home to place your cursor at the beginning of this sentence.

2. Press ctrl+RightArrow until you hear "brown"

3. Press ctrl+LeftArrow to move one word to the left and you will hear "quick"

4. Press ctrl+RightArrow and you will hear "brown" again

5. In the word saying "brown," use LeftArrow or RightArrow to move your cursor character-by-character

* Note: The expected behavior is that you hear each of the characters in "brown," but you will hear different characters as your cursor tracking is not mapped properly.

**Comment by aiday-mar** [comment]
Hi @jooyoungseo thanks for the steps, I have not tested this case. I will investigate the issue with these steps.

**Comment by aiday-mar** [comment]
Hi everyone, I have investigated the issue and I have traced this issue to the fact that we don't update the editor cursor when the EditContext dom node selection is changed. When you navigate from word to word, the selection changes inside of the dom node, and this has to be propagated to the editor. The following PR should fix this: https://github.com/microsoft/vscode/pull/230204.

**Comment by rperez030** [comment]
I have also identified an issue related to blank lines. When reading a file line by line with NVDA, it appears that NVDA does not correctly identify empty lines. I observed that the Unicode character generated by the Enter key varies depending on whether the setting is enabled.

With the setting disabled, the Unicode output is: "10,0 x a".

With the setting enabled, the Unicode output is: "13,0 x d".

NVDA should be able to identify new lines in both cases, so this is probably more related to the overall cursor tracking issue, but wanted to report it just in case.

**Comment by aiday-mar** [comment]
Ah yes I see, so in one case it is a Line Feed character and in the other case a Carriage Return character. I will look into this.

**Comment by aiday-mar** [comment]
Hi after some investigation, I discovered this was happening because of the recent change I did to fix the selection-change issue. My fix introduced a new issue in that the text for the screen readers is not correctly aligned with the cursor position. Essentially there was an offset issue, which I fix with https://github.com/microsoft/vscode/pull/230475.
